File: 231. Power of Two.cpp
class Solution {
public:
    // recursive: O(logN), 17.57%, ok
    bool isPowerOfTwo(int n) {
        if(n <= 0) return 0;
        if(n == 1) return true;
        return !(n % 2) && isPowerOfTwo(n / 2);
    }

    // iterative: O(logN), 17.57%, ok
    bool isPowerOfTwo(int n) {
        if(n <= 0) return 0;
        if(n == 1) return true;
        while(!(n % 2)) n /= 2;
        return n == 1;
    }

    // bit count: O(1), 17.57%, ok
    bool isPowerOfTwo(int n) {
        if(n <= 0) return 0;
        bitset<32> bit(n);
        return bit.count() == 1;
    }

    // bit manipulation: O(1), 17.57%, ok
    bool isPowerOfTwo(int n) {
        if(n <= 0) return 0;
        return !(n & (n - 1));
    }

    // kind of cheating: O(1), 17.57%, ok
    bool isPowerOfTwo(int n) {
        if(n <= 0) return 0;
        return !((int)pow(2, 30) % n);
    }
};

File: 231. Power of Two.py
class Solution(object):
    
    # recursion: O(logN) time, no space
    def isPowerOfTwo(self, n):
        """
        :type n: int
        :rtype: bool
        """
        if n <= 0: return False
        if n == 1: return True
        return n % 2 == 0 and self.isPowerOfTwo(n / 2)
    
    # iterative: O(logN) time, no space
    def isPowerOfTwo(self, n):
        if n <= 0: return False
        while n != 1:
            if n % 2 != 0: return False
            n /= 2
        return True
    
    # bit manipulation: O(1) time, no space
    def isPowerOfTwo(self, n):
        return n > 0 and n & (n-1) == 0

        